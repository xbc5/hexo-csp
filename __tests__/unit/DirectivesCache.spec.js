const DirectivesCache = require("../../lib/DirectivesCache");

const toAdd = (num) => ({
  base: `base${num}`,
  path: `path${num}`,
  canonical_path: `canonical_path${num}`,
  slug: `slug${num}`,
  permalink: `permalink${num}`,
  full_source: `full_source${num}`,
  directives: {
    [`directive${num}`]: [`source${num}a`, `source${num}b`],
  },
});

function fixture() {
  const log = { warn: jest.fn() };
  const file = { read: jest.fn(), write: jest.fn(), log };
  const fileData = '{ "foo": "bar" }';
  const readData = JSON.parse(fileData); // file data read, and stored in memory
  file.read.mockReturnValue(fileData);
  const cache = new DirectivesCache(file, log);
  return { cache, file, log, fileData, readData };
}

it("should call injected file.read when calling read()", async () => {
  const { cache, file } = fixture();
  cache.read();
  expect(file.read).toHaveBeenCalledTimes(1);
});

it("should populate directives without having to call read() first", async () => {
  const { cache, readData } = fixture();
  expect(cache.directives).toStrictEqual(readData);
});

describe("after adding data", () => {
  function run() {
    const { cache, file } = fixture();
    cache.add("key1", toAdd(1));
    cache.add("key2", toAdd(2));
    cache.write();
    return { cache, file };
  }

  it("should write it to file when instructed", async () => {
    const { file } = run();
    expect(file.write.mock.calls[0]).toMatchSnapshot();
  });

  it("should keep read data in sync with the written data", async () => {
    const { cache } = run();
    expect(cache.directives).toMatchSnapshot();
  });
});
