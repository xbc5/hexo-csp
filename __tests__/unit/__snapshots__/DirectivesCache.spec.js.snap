// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`after adding data should keep read data in sync with the written data 1`] = `
Object {
  "key1": Object {
    "base": "base1",
    "canonical_path": "canonical_path1",
    "directives": Object {
      "directive1": Array [
        "source1a",
        "source1b",
      ],
    },
    "full_source": "full_source1",
    "path": "path1",
    "permalink": "permalink1",
    "slug": "slug1",
  },
  "key2": Object {
    "base": "base2",
    "canonical_path": "canonical_path2",
    "directives": Object {
      "directive2": Array [
        "source2a",
        "source2b",
      ],
    },
    "full_source": "full_source2",
    "path": "path2",
    "permalink": "permalink2",
    "slug": "slug2",
  },
}
`;

exports[`after adding data should write it to file when instructed 1`] = `
Array [
  "{
  \\"key1\\": {
    \\"base\\": \\"base1\\",
    \\"path\\": \\"path1\\",
    \\"canonical_path\\": \\"canonical_path1\\",
    \\"slug\\": \\"slug1\\",
    \\"permalink\\": \\"permalink1\\",
    \\"full_source\\": \\"full_source1\\",
    \\"directives\\": {
      \\"directive1\\": [
        \\"source1a\\",
        \\"source1b\\"
      ]
    }
  },
  \\"key2\\": {
    \\"base\\": \\"base2\\",
    \\"path\\": \\"path2\\",
    \\"canonical_path\\": \\"canonical_path2\\",
    \\"slug\\": \\"slug2\\",
    \\"permalink\\": \\"permalink2\\",
    \\"full_source\\": \\"full_source2\\",
    \\"directives\\": {
      \\"directive2\\": [
        \\"source2a\\",
        \\"source2b\\"
      ]
    }
  }
}",
]
`;
